### ISSUE TITLE
Add more concrete error types

### ISSUE THREAD

**Issue by thomastaylor312** [nats-io/nats.rs#1200]
### Proposed change

The error types from the latest version have been a great improvement, but they still squash a lot of useful errors. Before this crate hits 1.0, I'd really like to add some additional error types that map to all the error codes or at least mark the error kind enums as non-exhaustive so things can be added in the future.

### Use case

There are several here, but one concrete thing I ran into recently was when creating a KV bucket. If I wanted to check if the bucket existed (via `get_key_value`) so I could log something to the user if it didn't already exist and then create the bucket, However, the error kind gets swallowed up in the `JetStream` error type so I can't check for something like `BucketDoesNotExist`. 

### Contribution

If I can make some time, I'd love to add at least some of the error kinds

**Comment by Jarema** [comment]
Thanks for raising this question.

There are few things to discuss here:

## Current errors
We purposely didn't expose all possible errors as separate enum variants, while wrapping the underlying details so they are still accesible. The goal was to expose those errors, against which users can do something.
I'm interested to hear what errors do you miss?
If you could give some examples, it would help.
Is it a rare case, or something you need to do often?
Does downcasting on `source` help?

## non-exhaustive errors
We were considering it. However I feel it affects the ergonomics of the client, so we made the call to not have them exhaustive.

## 1.0.0
I feel that in Rust ecosystem, especially when you are having <1.0 dependencies, it's a hard task to be on >1.0.
So for now, we decided to not land 1.0 technically, as some dependencies can introeduce breaking changes that will affect our API, like rustls which happens already few times.

Additionaly - technically speaking, expanding config structs of streams and consumer configs, or kv config, is a breaking change. And as you know - we do expand them every single Server non-patch release.

Instead, I think, a keeping the client on 0.x version with clear API guarantees is the way to go. The guarantees would be: we will expand struct configs and enum error varians when necessary, we might break TLS direct rustls configs (if rustls introduces breaking changes), but all those changes will be well documented, should be caught during compilation, easy to fix, with info how in each release whenever those happen.
That way we can also leverage some quality of life improvements happening in async ecosystem, like async traits, or hopefully coming this year - async closures, which would make `events` callback on connection so much nicer experience.

**Comment by Jarema** [comment]
I understand your point. However, you should be able to downcast then, rather than match string. I'll write example of that little later and will make sure it's the case. Anyway - it definately make sense to expand some of the enums. The idea is - to expose only those which make sense for user of the library.

**Comment by thomastaylor312** [comment]
What type do I downcast to? Are you talking about the `ErrorCode` type for jetstream?

**Comment by Jarema** [comment]
Yes, that is what I mean. It should be able to downcast to the `ErrorCode`, which handles all errors returned from the server.

**Comment by hseeberger** [comment]
So if I have an `async_nats::jetstream::kv::UpdateError`, which is an alias for `async_nats::error::Error`, how can I access the source to then downcast to a `async_nats::jetstream::Error` which gives me the error code?

**Comment by thomastaylor312** [comment]
Yeah, that is still a bit rough to do in practice because of all the error/options to handle. Curious what ideas you've had to make it better!

**Comment by m-hilgendorf** [comment]
I think having to do this is quite cumbersome, but necessary in order to handle some explicit errors like stream limits:

```rs
let code = error
    .source()
    .unwrap()
    .downcast_ref::<async_nats::jetstream::Error>()
    .unwrap()
    .code();
``` 

If you look at the [sqlx::Error](https://docs.rs/sqlx/latest/sqlx/enum.Error.html#method.as_database_error) type, they have a methods `as_database_error` or `into_data_base_error` which allow extracting the underlying protocol error and matching on the code. It would be useful for more advanced uses of the `Error` types in `async_nats` to be able to do this in a single line of code and avoid the unwraps and downcasting.
